code/src/reducers/labyrinth.js

```diff
@@ -0,0 +1,80 @@
+import { createSlice } from "@reduxjs/toolkit";
+
+const labyrinth = createSlice({
+    name: "labyrinth",
+    initialState: {
+        player: null,
+        stage: null,
+        history: [],
+        loading: false,
+    },
+    reducers: {
+        setPlayer: (store, action) => {
+            store.player = action.payload;
+        },
+        setStage: (store, action) => {
+            if (store.stage) {
+                store.history.push(store.stage)
+            }
+            store.stage = action.payload;
+
+        },
+        setPreviousStage: (store, action) => {
+            if (store.history.length !== 0) {
+                store.stage = store.history[store.history.length - 1]
+                store.history.splice(-1, 1)
+            }
+        },
+        setRestartGame: (store, action) => {
+            store.stage = null;
+            store.player = null;
+            store.history = [];
+        },
+        toggleLoading: (store, action) => {
+            console.log("toggle loading to", action.payload)
+            store.loading = action.payload
+        },

```

Comment:
Really nice reducers! Love that you guys stored the history and how you handled that. Big thumbs up!